File: java/sys_nat/fan/sys/FloatArray.java
package fan.sys;

import fanx.main.Sys;
import fanx.main.Type;

public abstract class FloatArray extends FanObj {

	public static FloatArray makeF4(long size) { return new F4((int)size); }
	public static FloatArray makeF8(long size) { return new F8((int)size); }

	public abstract double get(long pos);

	public abstract void set(long pos, double val);

	public abstract long size();

	public abstract Object raw();
	
	public FloatArray fill(double val, long times) {
		int t = (int) times;
		for (int i = 0; i < t; ++i) {
			set(i, val);
		}
		return this;
	}

	public abstract FloatArray realloc(long newSize);

	public abstract FloatArray copyFrom(FloatArray that, long thatOffset, long thisOffset, long length);

	public final Type typeof() {
		if (type == null)
			type = Sys.findType("sys::FloatArray");
		return type;
	}

	private static Type type;

	static class F4 extends FloatArray {
		final float[] array;

		public F4(long size) {
			array = new float[(int) size];
		}

		public Object raw() {
			return array;
		}

		public double get(long pos) {
			return array[(int) pos];
		}

		public void set(long pos, double val) {
			array[(int) pos] = (float) val;
		}

		public long size() {
			return array.length;
		}

		public FloatArray realloc(long newSize) {
			if (array.length == newSize) return this;
			F4 na = new F4(newSize);
			int len = array.length > na.array.length ? na.array.length : array.length;
			System.arraycopy(array, 0, na.array, 0, len);
			return na;
		}

		public F4 copyFrom(FloatArray that, long thatOffset, long thisOffset, long length) {
			System.arraycopy(((F4) that).array, (int) thatOffset, array, (int) thisOffset, (int) length);
			return this;
		}
	}

	static class F8 extends FloatArray {
		double[] array;

		public F8(long size) {
			array = new double[(int) size];
		}

		public Object raw() {
			return array;
		}

		public double get(long pos) {
			return array[(int) pos];
		}

		public void set(long pos, double val) {
			array[(int) pos] = val;
		}

		public long size() {
			return array.length;
		}

		public FloatArray realloc(long newSize) {
			if (array.length == newSize) return this;
			F8 na = new F8(newSize);
			int len = array.length > na.array.length ? na.array.length : array.length;
			System.arraycopy(array, 0, na.array, 0, len);
			return na;
		}

		public F8 copyFrom(FloatArray that, long thatOffset, long thisOffset, long length) {
			System.arraycopy(((F8) that).array, (int) thatOffset, array, (int) thisOffset, (int) length);
			return this;
		}
	}
}

File: src/sys/js/FloatArray.js


fan.sys.FloatArray = fan.sys.Obj.$extend(fan.sys.Obj);
fan.sys.FloatArray.prototype.$ctor = function() {}
fan.sys.FloatArray.prototype.$typeof = function() { return fan.sys.FloatArray.$type; }

fan.sys.FloatArray.makeF4(long size) {
	self = new fan.sys.FloatArray()
	self.m_array = new Float32Array(size);
	return self
}

fan.sys.FloatArray.makeF8(long size) {
	self = new fan.sys.FloatArray()
	self.m_array = new Float64Array(size);
	return self
}

fan.sys.FloatArray.prototype.get = function(pos) {
	return this.m_array[pos]
}

fan.sys.FloatArray.prototype.set = function(pos, val) {
	this.m_array[pos] = val
}

fan.sys.FloatArray.prototype.size = function() {
	return this.m_size;
}

fan.sys.FloatArray.prototype.realloc = function(newSize) {
	if (this.m_array.length == newSize) return this;

	if (newSize > this.m_array.length) {
		for (i = this.m_array.length; i<newSize; ++i) {
			this.m_array.push(null)
		}
		return this;
	}

	na = fan.sys.FloatArray.make(newSize, of);
	len = this.m_array.length > newSize ? newSize : this.m_array.length;
	for (i = 0; i<len; ++i) {
		na.m_array[i] = this.m_array[i]
	}
	return na;
}

fan.sys.FloatArray.prototype.fill = function(val, times) {
	for (i = 0; i < times; ++i) {
		this.m_array[i] = val;
	}
	return this;
}

fan.sys.FloatArray.prototype.copyFrom = function(that, thatOffset, thisOffset, length) {
	if (this == that) {
		this.m_array.copyWithin(thisOffset, thisOffset, thisOffset+length)
		return this;
	}

	for (i = 0; i<length; ++i) {
		this.m_array[thisOffset + i] = that.m_array[i+thatOffset]
	}
	return this;
}

